### DISCUSSION TITLE
Supporting Aggregate and other operators

### DISCUSSION THREAD

**Discussion by psicomante** [maxnowack/signaldb#924] [General]
Hello Max,

I have another question about operators supported in Mingo. Am i wrong or aggregate selectors like:

```js
[
  {
    $unwind: "$complex"
  },
  {
    $match: {
      "complex.type": 4
    }
  },
  {
    $project: {
      _id: 0,
      id: "$complex.id",
      type: "$complex.type"
    }
  }
]
```

are not yet supported?

If there is no way to use this type of selector inside Signaldb, do you plan to add support for it?

**Comment by maxnowack** [comment]
Hi @psicomante, currently SignalDB doesn't support the aggregation framework, as it would add too much complexity and also is a bit out of scope of the core functionality. Maybe I'll add support some day, but it's not planned right now.
However, you should be able to use the aggregation functionality of `mingo` directly.

Example (not tested, but should work kind of like this):
```js
import { Collection } from 'signaldb'
import { Aggregator } from 'mingo/aggregator'
import { useOperators, OperatorType } from 'mingo/core'
import { $match, $group } from 'mingo/operators/pipeline'
import { $min } from 'mingo/operators/accumulator'

// ensure the required operators are preloaded prior to using them.
useOperators(OperatorType.PIPELINE, { $match, $group })
useOperators(OperatorType.ACCUMULATOR, { $min })

const agg = new Aggregator([
  { $match: { type: 'homework' } },
  { $group: { _id: '$student_id', score: { $min: '$score' } } },
  { $sort: { _id: 1, score: 1 } },
])

const collection = new Collection()
const result = agg.run(collection.find().fetch())
```

**Reply by psicomante** [reply]
Thanks! I will give a try! The fundamental question is how to manage complex data with external (kind of foreign) keys.

**Reply by maxnowack** [reply]
I manage those data by transforming items of the collection to a class instance that have methods to derive relations. I'll have to write a guide for this some day 😀 
Here is a very short example:

```js
class Post {
  constructor(item) {
    Object.assign(this, item)
  }

  getAuthor() {
    return Authors.findOne({ id: this.authorId })
  }

  getComments() {
    return Comments.find({ postId: this.id }).fetch()
  }
}

const Posts = new Collection({
  transform: item => new Post(item), 
})

const posts = Posts.find().map(post => ({
  title: post.title,
  authorName: post.getAuthor().name,
  comments: post.getComments(),
}))
```

**Reply by maxnowack** [reply]
I wrote some more detailed info on this here: https://signaldb.js.org/orm/
